### ISSUE TITLE
discovery when using device_identity fails if device has multiple device_ip, but the desired one is already set as canonical address

### ISSUE THREAD

**Issue by sjuerges** [netdisco/netdisco#1200]
## Context
We had a problem that devices that have multiple device_ips were using the wrong one as the canonical address. 
We are currently trying to fix this by utilizing the device_identity parameter (or, alternatively, renumber) 

Both are not working, because not ALL devices use the wrong address, but the discovery of devices that use the desired address as canonical fails:

Error Output:
```
info discover: 
status error: DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: ERROR:  duplicate key value violates unique constraint "device_ip_pkey"
DETAIL:  Key (ip, alias)=(172.19.16.190, 172.19.16.190) already exists. [for Statement "UPDATE device_ip SET ip = ? WHERE ( ip = ? )" with ParamValues: 1='172.19.16.190', 2='10.1.108.190'] at /data/netdisco/perl5/lib/perl5/App/Netdisco/DB/Result/Device.pm line 350 
```

used device_identity stanza:
`device_identity: 'vendor:cisco': '172.19.16.0/23'`

## Expected Behavior
using device_identity on devices with multiple device_ip ensures that the correct ip is used as the canonical address. an update is only issued if the WRONG address is being used as canoncial

## Current Behavior
when using device_identity, netdisco ALWAYS tries to set the canonical ip to the selected one IF an address from the desired ACL is found, without checking if that address is already set as the canonical ip.

## Possible Solution
integrate check if new_device_ip == old_device ip, skip update if true

## Steps to Reproduce (for bugs)
1. have two devices, both with at least two device_ip adresses
2. discovery devices (first without device_identity)
3. one device with canonical ip from first subnet, other devices with canonical ip from other subnet
4. set device_identity stanza, so that both devices should use the same subnet for canonical address
5. start discovery that would touch both devices
6. discovery fails with the aforementioned error when trying to set the new canonical_ip on the device that already has the canonical ip from the desired subnet

## Your Environment (updated 05.06.2024)
App::Netdisco 	2.76.5
SNMP::Info 	3.970.1
DB Schema 	87
PostgreSQL 	14.00.9
Perl 	5.34.0

## Config info (deployment.yml)
relevant change: 
device_identity: 'vendor:cisco': '172.19.16.0/23'

**Comment by ollyg** [comment]
Hi @sjuerges many thanks for the bug report! I believe this is fixed in d8c850a0 and I will release 2.075003 now

**Comment by ollyg** [comment]
OK I've reopened the issue, thanks for the update

**Comment by sjuerges** [comment]
@ollyg In case you need more information or you want me to check things / run commands, I am available via teams as well. Just let me know :)

**Comment by njohnsn** [comment]
What's weird is that if I go into the Postgres database itself, to the "device_ip" table, and try to search for the key (ip, alias) or even just the ip or alias individually, there is nothing there.

If in insert a row with the information that is in the exception, it is added fine. If I then rerun the discovery for the device, it bombs out another attempt to insert.

**Comment by njohnsn** [comment]
What's weird is that if I go into the Postgres database itself, to the "device_ip" table, and try to search for the key (ip, alias) or even just the ip or alias individually, there is nothing there.

if in insert a row with the information that is in the exception, it is added fine. If I then rerun the discovery for the device, it bombs out another attempt to insert.

Trying to find in the source code where the actual insert into the database is made

**Comment by njohnsn** [comment]
What's weird is that if I go into the Postgres database itself, to the "device_ip" table, and try to search for the key (ip, alias) or even just the ip or alias individually, there is nothing there.

If I insert a row into the "device_ip" table with the information that is in the exception, it is added fine. If I then rerun the discovery for the device, it bombs out a different insert.

Trying to find in the source code where the actually insert into the database is made.

**Comment by njohnsn** [comment]
I wish I could be more helpful, but my PERL skills don't include objects or ORMs. :-)

**Comment by ollyg** [comment]
Hi @njohnsn I think your error is a different one, but is fixed via ticket #1232 that I have just resolved for the next release.

I'll still take a look into the issue reported by @sjuerges

**Comment by ollyg** [comment]
Hi @sjuerges would you indulge me by testing again? I am pretty sure the fix I already released in version 2.75.3

Indeed `renumber` will NOT work for this, as the last line of the output says: "[882089] 2024-06-05 09:31:38  info renumber: status error: already know new device as: 10.1.108.111 (use device_identity instead)."

But `device_identity` should work fine now. So you can repeat your test as above:

1. have two devices, both with at least two device_ip adresses
2. discovery devices (first without device_identity)
3. one device with canonical ip from first subnet, other devices with canonical ip from other subnet
4. set device_identity stanza, so that both devices should use the same subnet for canonical address
5. start discovery that would touch both devices
6. discovery fails with the aforementioned error when trying to set the new canonical_ip on the device that already has the canonical ip from the desired subnet

**Comment by sjuerges** [comment]
Hey @ollyg, I'm now back from summer holidays and will get to testing. Sorry for the delay

**Comment by sjuerges** [comment]
Hey, unfortunately, no dice:

Error: (also attaching full DISQ Output
[full_disq_debug_output.txt](https://github.com/user-attachments/files/16928102/full_disq_debug_output.txt)

```
[182461] 2024-09-09 08:51:24  info discover: finished at Mon Sep  9 10:51:24 2024
[182461] 2024-09-09 08:51:24  info discover: status error: DBI Exception: DBD::Pg::st execute failed: ERROR:  duplicate key value violates unique constraint "device_module_pkey"
DETAIL:  Key (ip, index)=(172.19.16.68, 2077) already exists. [for Statement "UPDATE device_module SET ip = ? WHERE ( ip = ? )" with ParamValues: 1='172.19.16.68', 2='10.1.108.111'] at /data/netdisco/perl5/lib/perl5/DBIx/Class/Schema.pm line 1118.
	DBIx::Class::Schema::throw_exception(DBIx::Class::Version=HASH(0x55fd6e130100), "DBI Exception: DBD::Pg::st execute failed: ERROR:  duplicate "...) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage.pm line 113
	DBIx::Class::Storage::throw_exception(DBIx::Class::Storage::DBI::Pg=HASH(0x55fd6e12ff68), "DBI Exception: DBD::Pg::st execute failed: ERROR:  duplicate "...) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/DBI.pm line 1623
	DBIx::Class::Storage::DBI::__ANON__("DBD::Pg::st execute failed: ERROR:  duplicate key value viola"..., DBI::st=HASH(0x55fd702a8798), undef) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/DBI.pm line 1958
	DBIx::Class::Storage::DBI::_dbh_execute(DBIx::Class::Storage::DBI::Pg=HASH(0x55fd6d49b770), DBI::db=HASH(0x55fd6e3783b8), "UPDATE device_module SET ip = ? WHERE ( ip = ? )", ARRAY(0x55fd702e53a8), ARRAY(0x55fd702af118)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/DBI.pm line 839
	DBIx::Class::Storage::DBI::dbh_do(DBIx::Class::Storage::DBI::Pg=HASH(0x55fd6d49b770), "_dbh_execute", "UPDATE device_module SET ip = ? WHERE ( ip = ? )", ARRAY(0x55fd702e53a8), ARRAY(0x55fd702af118)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/DBI.pm line 1939
	DBIx::Class::Storage::DBI::_execute(DBIx::Class::Storage::DBI::Pg=HASH(0x55fd6d49b770), "update", DBIx::Class::ResultSource::Table=HASH(0x55fd6de72328), HASH(0x55fd702de188), REF(0x55fd702af3d0)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/DBI.pm line 2573
	DBIx::Class::Storage::DBI::update(DBIx::Class::Storage::DBI::Pg=HASH(0x55fd6d49b770), DBIx::Class::ResultSource::Table=HASH(0x55fd6de72328), HASH(0x55fd702de188), REF(0x55fd702af3d0)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/ResultSet.pm line 2009
	DBIx::Class::ResultSet::_rs_update_delete(App::Netdisco::DB::ResultSet::DeviceModule=HASH(0x55fd701c99b8), "update", HASH(0x55fd702de188)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/ResultSet.pm line 2058
	DBIx::Class::ResultSet::update(App::Netdisco::DB::ResultSet::DeviceModule=HASH(0x55fd701c99b8), HASH(0x55fd702de188)) called at /data/netdisco/perl5/lib/perl5/App/Netdisco/DB/Result/Device.pm line 352
	App::Netdisco::DB::Result::Device::renumber(App::Netdisco::DB::Result::Device=HASH(0x55fd6e7457d0), "172.19.16.68") called at /data/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Plugin/Discover/CanonicalIP.pm line 89
	App::Netdisco::Worker::Plugin::Discover::CanonicalIP::__ANON__() called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/BlockRunner.pm line 130
	DBIx::Class::Storage::BlockRunner::try {...} () called at /usr/share/perl5/Try/Tiny.pm line 98
	eval {...} called at /usr/share/perl5/Try/Tiny.pm line 93
	Try::Tiny::try(CODE(0x55fd701ca4c8), Try::Tiny::Catch=REF(0x55fd702acba8)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/BlockRunner.pm line 134
	DBIx::Class::Storage::BlockRunner::__ANON__() called at /data/netdisco/perl5/lib/perl5/Context/Preserve.pm line 33
	Context::Preserve::preserve_context(CODE(0x55fd701b0a10), "replace", CODE(0x55fd7021e9e0)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/BlockRunner.pm line 213
	DBIx::Class::Storage::BlockRunner::_run(DBIx::Class::Storage::BlockRunner=HASH(0x55fd70215c78), CODE(0x55fd6f7156f8)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/BlockRunner.pm line 105
	DBIx::Class::Storage::BlockRunner::run(DBIx::Class::Storage::BlockRunner=HASH(0x55fd70215c78), CODE(0x55fd6f7156f8)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage.pm line 187
	DBIx::Class::Storage::txn_do(DBIx::Class::Storage::DBI::Pg=HASH(0x55fd6d49b770), CODE(0x55fd6f7156f8)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Storage/DBI.pm line 861
	DBIx::Class::Storage::DBI::txn_do(DBIx::Class::Storage::DBI::Pg=HASH(0x55fd6d49b770), CODE(0x55fd6f7156f8)) called at /data/netdisco/perl5/lib/perl5/DBIx/Class/Schema.pm line 654
	DBIx::Class::Schema::txn_do(App::Netdisco::DB=HASH(0x55fd6d47af90), CODE(0x55fd6f7156f8)) called at /data/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Plugin/Discover/CanonicalIP.pm line 98
	App::Netdisco::Worker::Plugin::Discover::CanonicalIP::__ANON__(App::Netdisco::Backend::Job=HASH(0x55fd6d487320), HASH(0x55fd6eb0cc68)) called at /data/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Plugin.pm line 117
	App::Netdisco::Worker::Plugin::__ANON__(App::Netdisco::Backend::Job=HASH(0x55fd6d487320)) called at /data/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Runner.pm line 105
	App::Netdisco::Worker::Runner::try {...} () called at /usr/share/perl5/Try/Tiny.pm line 102
	eval {...} called at /usr/share/perl5/Try/Tiny.pm line 93
	Try::Tiny::try(CODE(0x55fd6f6520b8), Try::Tiny::Catch=REF(0x55fd7039fb38)) called at /data/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Runner.pm line 109
	App::Netdisco::Worker::Runner::run_workers(MyWorker=HASH(0x55fd6d44f540), "workers_main") called at /data/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Runner.pm line 74
	App::Netdisco::Worker::Runner::__ANON__(MyWorker=HASH(0x55fd6d44f540), App::Netdisco::Backend::Job=HASH(0x55fd6d487320)) called at /data/netdisco/perl5/lib/perl5/Sys/SigAction.pm line 164
	eval {...} called at /data/netdisco/perl5/lib/perl5/Sys/SigAction.pm line 162
	eval {...} called at /data/netdisco/perl5/lib/perl5/Sys/SigAction.pm line 160
	Sys::SigAction::timeout_call(600, CODE(0x55fd6d417138), MyWorker=HASH(0x55fd6d44f540), App::Netdisco::Backend::Job=HASH(0x55fd6d487320)) called at /data/netdisco/perl5/lib/perl5/App/Netdisco/Worker/Runner.pm line 82
	App::Netdisco::Worker::Runner::run(MyWorker=HASH(0x55fd6d44f540), App::Netdisco::Backend::Job=HASH(0x55fd6d487320)) called at /data/netdisco/bin/netdisco-do line 195
	main::try {...} () called at /usr/share/perl5/Try/Tiny.pm line 102
	eval {...} called at /usr/share/perl5/Try/Tiny.pm line 93
	Try::Tiny::try(CODE(0x55fd6d47ac60), Try::Tiny::Catch=REF(0x55fd6bb5cf60)) called at /data/netdisco/bin/netdisco-do line 200
```

**Comment by ollyg** [comment]
Hi again @sjuerges, my best idea at this error is that you have some cruft left in the database which is tripping up Netdisco when it does the renumber during discover. Basically some stuff for device 172.19.16.68 left over. This might work and then run discover again:

`~/bin/netdisco-do psql -e 'DELETE FROM device_module WHERE ip = "172.19.16.68"'`

Alternatively "netdisco-do delete" both device IPs and then rediscover from scratch as 172.19.16.68.

**Comment by sjuerges** [comment]
Hey @ollyg, thanks for getting back to me. I tried the first decribed way (psql), but that just kept showing more tables. Apparently, log of cruft. So I ended up deleting devices, wanting to do a rediscover. That ended up showing me the issue that my devices were reporting the wrong IP in CDP, leading to autodiscovery failures (it only worked coz the old ips were still in the DB)

fixing that issues (using 
`cdp mgmt-address ipv4 <MYIP>`
lead to postive discovery on my test device, and correct setting of the canonical address.

This was a bit of a layered problem, so thanks for your assitance. 
It turn out that 
a) having a really old db (as in, imported from nd1) and
b) triggering a bug somewhere along the line that got already fixed and
c) having cdp behaving weirdly 

is a bad combination.
From my end, this is solved. Lots of work in my future, but solved ;)
